### PR TITLE
Support restore graph in restoring and merging mode

### PR DESCRIPTION
**Pull request by zhoney**
implement: #182 

Change-Id: I7c2c85435af01c5d27279076d0473685b8f23f34

### CODE REVIEW

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/BackendStore.java [hunk 1]
```diff
@@ -63,7 +65,47 @@
...
+
+        E.checkState(counter != 0L, "Please check whether '%s' is OK",
+                     this.provider().type());
+        E.checkState(counter == expect, "'%s' is busy please try again",
+                     this.provider().type());
+        return IdGenerator.of(expect);
+    }
+
+    // Set next id >= lowest for a specific type
+    public default void setCounter(HugeType type, long lowest) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/BackendStore.java]: setCounterLowest()

File: hugegraph-hbase/src/main/java/com/baidu/hugegraph/backend/store/hbase/HbaseTables.java [hunk 2]
```diff
@@ -54,10 +54,21 @@ public Counters() {
...
 
-        public synchronized Id nextId(Session session, HugeType type) {
+        public long getCounter(Session session, HugeType type) {
             byte[] key = new byte[]{type.code()};
-            long counter = session.increase(this.table(), CF, key, COL, 1L);
-            return IdGenerator.of(counter);
+            RowIterator results = session.get(this.table(), CF, key);
+            if (results.hasNext()) {
+                Result row = results.next();
+                return ByteBuffer.wrap(row.getValue(CF, COL)).getLong();
```
Comment (javeme) [inline hugegraph-hbase/src/main/java/com/baidu/hugegraph/backend/store/hbase/HbaseTables.java]: call NumericUtil.bytesToLong(byte[] bytes)

File: hugegraph-core/src/main/java/com/baidu/hugegraph/schema/builder/VertexLabelBuilder.java [hunk 3]
```diff
@@ -103,23 +112,20 @@ public VertexLabel build() {
 
     @Override
     public VertexLabel create() {
-        SchemaElement.checkName(this.name,
-                                this.transaction.graph().configuration());
-        VertexLabel vertexLabel = this.transaction.getVertexLabel(this.name);
+        SchemaTransaction tx = this.transaction;
+        SchemaElement.checkName(this.name, tx.graph().configuration());
+        VertexLabel vertexLabel = (VertexLabel) tx.checkExist(
+                    HugeType.VERTEX_LABEL, this.name, this.id, this.checkExist);
```
[the remark this answers is not in the document]
Comment (zhoney) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/schema/builder/VertexLabelBuilder.java]: already align with "vertexLabel"

File: hugegraph-test/src/main/java/com/baidu/hugegraph/core/RestoreCoreTest.java [hunk 4]
```diff
@@ -0,0 +1,731 @@
...
+        graph.mode(GraphMode.RESTORING);
+        Assert.assertThrows(UnsupportedOperationException.class, () ->
+            v1.addEdge("knows", v2, T.id, "id")
+        );
+    }
+
+
+
+
+
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/core/RestoreCoreTest.java]: empty lines

File: hugegraph-core/src/main/java/com/baidu/hugegraph/type/HugeType.java [hunk 5]
```diff
@@ -53,6 +53,9 @@
 
     TASK(180, "T"),
 
+    // System schema
+    SYS_SCHEMA(190, "SS"),
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/type/HugeType.java]: set to 250

File: hugegraph-core/src/main/java/com/baidu/hugegraph/schema/builder/PropertyKeyBuilder.java [hunk 6]
```diff
@@ -59,8 +61,13 @@ public PropertyKeyBuilder(String name, SchemaTransaction transaction) {
 
     @Override
     public PropertyKey build() {
+        Id id = this.transaction.generateId(HugeType.PROPERTY_KEY, this.id,
+                                            this.name);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/schema/builder/PropertyKeyBuilder.java]: put this.id on the next line

File: hugegraph-core/src/main/java/com/baidu/hugegraph/schema/builder/IndexLabelBuilder.java [hunk 7]
```diff
@@ -91,7 +97,8 @@ public IndexLabel build() {
     public IndexLabel.CreatedIndexLabel createWithTask() {
         SchemaElement.checkName(this.name,
                                 this.transaction.graph().configuration());
-        IndexLabel indexLabel = this.transaction.getIndexLabel(this.name);
+        IndexLabel indexLabel = (IndexLabel) this.transaction.checkExist(
+                   HugeType.INDEX_LABEL, this.name, this.id, this.checkExist);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/schema/builder/IndexLabelBuilder.java]: align

File: hugegraph-core/src/main/java/com/baidu/hugegraph/schema/builder/EdgeLabelBuilder.java [hunk 8]
```diff
@@ -109,17 +117,13 @@ public EdgeLabel build() {
...
     @Override
     public EdgeLabel create() {
-        SchemaElement.checkName(this.name,
-                                this.transaction.graph().configuration());
-
-        EdgeLabel edgeLabel = this.transaction.getEdgeLabel(this.name);
+        SchemaTransaction tx = this.transaction;
+        SchemaElement.checkName(this.name, tx.graph().configuration());
+        EdgeLabel edgeLabel = (EdgeLabel) tx.checkExist(
+                  HugeType.EDGE_LABEL, this.name, this.id, this.checkExist);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/schema/builder/EdgeLabelBuilder.java]: align

File: hugegraph-core/src/main/java/com/baidu/hugegraph/HugeGraph.java [hunk 9]
```diff
@@ -161,12 +163,17 @@ public boolean closed() {
...
+        return this.mode;
+    }
+
+    public void mode(GraphMode mode) {
+        this.mode = mode;
     }
 
-    public void restoring(boolean restoring) {
-        this.restoring = restoring;
+    public boolean restoring() {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/HugeGraph.java]: it's not used any more?

File: hugegraph-core/src/main/java/com/baidu/hugegraph/schema/builder/VertexLabelBuilder.java [hunk 10]
```diff
@@ -190,6 +195,14 @@ public Id rebuildIndex() {
         return this.transaction.rebuildIndex(vertexLabel);
     }
 
+    @Override
+    public VertexLabelBuilder id(long id) {
+        E.checkArgument(id != 0L,
+                        "Not allowed to assign 0 as vertex label id");
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/schema/builder/VertexLabelBuilder.java]: put on the same line?

File: hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBTables.java [hunk 11]
```diff
@@ -41,38 +41,24 @@
...
+                return 0L;
             }
+        }
 
-            E.checkState(counter != 0L, "Please check whether RocksDB is OK");
-            E.checkState(counter == expect, "RocksDB is busy please try again");
-            return IdGenerator.of(counter);
+        public void increaseCounter(Session session, HugeType type,
+                                    long increment) {
+            session.increase(this.table(), new byte[]{type.code()},
```
Comment (javeme) [inline hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBTables.java]: define key:
byte[] key = new byte[]{type.code()};

File: hugegraph-core/src/main/java/com/baidu/hugegraph/schema/builder/VertexLabelBuilder.java (hunk 10 above)
Comment (zhoney) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/schema/builder/VertexLabelBuilder.java]: can't put on one line

File: hugegraph-core/src/main/java/com/baidu/hugegraph/HugeGraph.java (hunk 9 above)
Comment (zhoney) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/HugeGraph.java]: still used in constructVertex() and checkId() of GraphTransaction

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/BackendStore.java [hunk 12]
```diff
@@ -63,7 +65,47 @@
...
+    public default void setCounterLowest(HugeType type, long lowest) {
+        long current = this.getCounter(type);
+        if (current >= lowest) {
+            return;
+        }
+        long increment = lowest - current;
+        this.increaseCounter(type, increment);
+    }
+
+    // Set next id >= lowest for specific type
```
Comment (Linary) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/BackendStore.java]: check this comment

File: hugegraph-core/src/main/java/com/baidu/hugegraph/schema/builder/IndexLabelBuilder.java [hunk 13]
```diff
@@ -172,6 +175,14 @@ public Id rebuild() {
         return this.transaction.rebuildIndex(indexLabel);
     }
 
+    @Override
+    public IndexLabelBuilder id(long id) {
+        E.checkArgument(id != 0L,
+                        "Not allowed to assign 0 as index label id");
```
Comment (Linary) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/schema/builder/IndexLabelBuilder.java]: seems can be placed in one line

File: hugegraph-core/src/main/java/com/baidu/hugegraph/schema/builder/PropertyKeyBuilder.java [hunk 14]
```diff
@@ -132,6 +138,14 @@ public Id remove() {
         return null;
     }
 
+    @Override
+    public PropertyKeyBuilder id(long id) {
+        E.checkArgument(id != 0L,
```
Comment (Linary) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/schema/builder/PropertyKeyBuilder.java]: ditto

File: hugegraph-core/src/main/java/com/baidu/hugegraph/schema/builder/VertexLabelBuilder.java (hunk 10 above)
Comment (Linary) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/schema/builder/VertexLabelBuilder.java]: Just 80 characters

File: hugegraph-core/src/main/java/com/baidu/hugegraph/schema/builder/IndexLabelBuilder.java [hunk 15]
```diff
@@ -89,15 +91,16 @@ public IndexLabel build() {
...
-        SchemaElement.checkName(this.name,
-                                this.transaction.graph().configuration());
-        IndexLabel indexLabel = this.transaction.getIndexLabel(this.name);
+        SchemaTransaction tx = this.transaction;
+        SchemaElement.checkName(this.name, tx.graph().configuration());
+        IndexLabel indexLabel = tx.getIndexLabel(this.name);
         if (indexLabel != null) {
             if (this.checkExist) {
-                throw new ExistedException("index label", this.name);
+                throw new ExistedException("edge label", this.name);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/schema/builder/IndexLabelBuilder.java]: index label

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/SchemaTransaction.java [hunk 16]
```diff
@@ -361,6 +377,47 @@ private BackendEntry serialize(SchemaElement schema) {
...
+        E.checkState(graph.mode() == GraphMode.RESTORING,
+                     "Can't build schema with provided id '%s' " +
+                     "when graph '%s' in mode '%s'",
+                     id, graph, graph.mode());
+        this.setNextIdLowest(type, id.asLong());
+    }
+
+    public void checkIdIfRestoringMode(HugeType type, Id id) {
+        if (this.graph().mode() == GraphMode.RESTORING) {
+            E.checkNotNull(id, "id");
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/SchemaTransaction.java]: check arg not null

File: hugegraph-core/src/main/java/com/baidu/hugegraph/schema/builder/IndexLabelBuilder.java (hunk 13 above)
Comment (zhoney) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/schema/builder/IndexLabelBuilder.java]: can't put in one line

File: hugegraph-core/src/main/java/com/baidu/hugegraph/schema/builder/PropertyKeyBuilder.java (hunk 14 above)
Comment (zhoney) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/schema/builder/PropertyKeyBuilder.java]: ditto

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/SchemaTransaction.java [hunk 17]
```diff
@@ -361,6 +377,47 @@ private BackendEntry serialize(SchemaElement schema) {
         }
     }
 
+    public Id validOrGenerateId(HugeType type, Id id, String name) {
+        if (id != null) {
+            this.checkIdAndUpdateNextId(type, id, name);
+        } else {
+            if (Graph.Hidden.isHidden(name)) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/SchemaTransaction.java]: boolean forSystem = Graph.Hidden.isHidden(name)
and pass into checkIdAndUpdateNextId()

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/schema/EdgeLabelAPI.java [hunk 18]
```diff
@@ -178,6 +179,16 @@ public void checkCreate(boolean isBatch) {
 
         private EdgeLabel.Builder convert2Builder(HugeGraph g) {
             EdgeLabel.Builder builder = g.schema().edgeLabel(this.name);
+            if (this.id != 0) {
+                E.checkArgument(this.id > 0,
+                                "Only positive number can be assign as " +
+                                "edge label id");
+                E.checkState(g.mode() == GraphMode.RESTORING,
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/schema/EdgeLabelAPI.java]: checkArgument

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/schema/IndexLabelAPI.java [hunk 19]
```diff
@@ -153,6 +154,16 @@ public void checkCreate(boolean isBatch) {
 
         private IndexLabel.Builder convert2Builder(HugeGraph g) {
             IndexLabel.Builder builder = g.schema().indexLabel(this.name);
+            if (this.id != 0) {
+                E.checkArgument(this.id > 0,
+                                "Only positive number can be assign as " +
+                                "index label id");
+                E.checkState(g.mode() == GraphMode.RESTORING,
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/schema/IndexLabelAPI.java]: ditto

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/schema/PropertyKeyAPI.java [hunk 20]
```diff
@@ -173,6 +174,16 @@ public void checkCreate(boolean isBatch) {
 
         private PropertyKey.Builder convert2Builder(HugeGraph g) {
             PropertyKey.Builder builder = g.schema().propertyKey(this.name);
+            if (this.id != 0) {
+                E.checkArgument(this.id > 0,
+                                "Only positive number can be assign as " +
+                                "property key id");
+                E.checkState(g.mode() == GraphMode.RESTORING,
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/schema/PropertyKeyAPI.java]: ditto

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/schema/VertexLabelAPI.java [hunk 21]
```diff
@@ -177,6 +178,16 @@ public void checkCreate(boolean isBatch) {
 
         private VertexLabel.Builder convert2Builder(HugeGraph g) {
             VertexLabel.Builder builder = g.schema().vertexLabel(this.name);
+            if (this.id != 0) {
+                E.checkArgument(this.id > 0,
+                                "Only positive number can be assign as " +
+                                "vertex label id");
+                E.checkState(g.mode() == GraphMode.RESTORING,
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/schema/VertexLabelAPI.java]: ditto

File: hugegraph-hbase/src/main/java/com/baidu/hugegraph/backend/store/hbase/HbaseTables.java [hunk 22]
```diff
@@ -54,10 +54,21 @@ public Counters() {
...
+                Result row = results.next();
+                return NumericUtil.bytesToLong(row.getValue(CF, COL));
+            } else {
+                return 0L;
+            }
+        }
+
+        public void increaseCounter(Session session, HugeType type,
+                                    long increment) {
+            session.increase(this.table(), CF, new byte[]{type.code()},
```
Comment (javeme) [inline hugegraph-hbase/src/main/java/com/baidu/hugegraph/backend/store/hbase/HbaseTables.java]: byte[] key = new byte[]{type.code()};

**Pull request merged**